In [538]:
# Import the necessary libraries

import numpy as np
import pandas as pd
import tensorflow as tf
import os

# to make this notebook's output stable across runs
np.random.seed(123)

# To plot pretty figures
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
plt.rcParams['axes.labelsize'] = 14
plt.rcParams['xtick.labelsize'] = 12
plt.rcParams['ytick.labelsize'] = 12

In [1033]:
#####################################
# Import the train and test datasets#
#####################################

train_data_full = pd.read_csv("/home/isaac/Fundamentals_of_Data_Science_Certificate/Extra Python work/Titanic Project/train.csv")

test_data = pd.read_csv("/home/isaac/Fundamentals_of_Data_Science_Certificate/Extra Python work/Titanic Project/test.csv")

In [1034]:
train_data_full

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
..           ...       ...     ...   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN

In [1035]:
train_data_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


In [1036]:
train_data_full.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [1037]:
col_names = list(train_data_full)

for i in range(len(col_names)):
    col_name = col_names[i]
    print(col_name)
    string_data = str(col_name)
    print(train_data_full[string_data].value_counts())
    print(" ")

PassengerId
891    1
293    1
304    1
303    1
302    1
301    1
300    1
299    1
298    1
297    1
296    1
295    1
294    1
292    1
306    1
291    1
290    1
289    1
288    1
287    1
286    1
285    1
284    1
283    1
282    1
281    1
305    1
307    1
279    1
321    1
      ..
561    1
560    1
584    1
585    1
586    1
587    1
610    1
609    1
608    1
607    1
606    1
605    1
604    1
603    1
602    1
601    1
600    1
599    1
598    1
597    1
596    1
595    1
594    1
593    1
592    1
591    1
590    1
589    1
588    1
1      1
Name: PassengerId, Length: 891, dtype: int64
 
Survived
0    549
1    342
Name: Survived, dtype: int64
 
Pclass
3    491
1    216
2    184
Name: Pclass, dtype: int64
 
Name
Nosworthy, Mr. Richard Cater                     1
Chronopoulos, Mr. Apostolos                      1
Lahoud, Mr. Sarkis                               1
Mockler, Miss. Helen Mary "Ellie"                1
Hunt, Mr. George Henry                           1
Skoog, Mast

# Separate Labels and Features

- Train Data

In [1038]:
train_data = train_data_full.drop("Survived", axis = 1)
train_labels = train_data_full["Survived"].copy()

In [1039]:
train_data.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean    446.000000    2.308642   29.699118    0.523008    0.381594   32.204208
std     257.353842    0.836071   14.526497    1.102743    0.806057   49.693429
min       1.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%     223.500000    2.000000   20.125000    0.000000    0.000000    7.910400
50%     446.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%     668.500000    3.000000   38.000000    1.000000    0.000000   31.000000
max     891.000000    3.000000   80.000000    8.000000    6.000000  512.329200

# Drop the PassengerID

- Train Data

In [1040]:
passenger_id_train = train_data["PassengerId"].copy()
train_data = train_data.drop("PassengerId", axis = 1)

- Test Data

In [1041]:
passenger_id_test = test_data["PassengerId"].copy()
test_data = test_data.drop("PassengerId", axis = 1)

# Do some nonsense feature engineering

## Drop "Embarked"

- Train Data

In [1042]:
#passenger_id_train = train_data["Embarked"].copy()
#train_data = train_data.drop("Embarked", axis = 1)

- Test Data

In [1043]:
#passenger_id_train = test_data["Embarked"].copy()
#test_data = test_data.drop("Embarked", axis = 1)

## Drop "Ticket"

- Train Data

In [1044]:
ticket_train = train_data["Ticket"].copy()
train_data = train_data.drop("Ticket", axis = 1)

- Test Data

In [1045]:
ticket_test = test_data["Ticket"].copy()
test_data = test_data.drop("Ticket", axis = 1)

## Drop "Cabin"

- Train Data

In [1046]:
cabin_train = train_data["Cabin"].copy()
train_data = train_data.drop("Cabin", axis = 1)

- Test Data

In [1047]:
cabin_test = test_data["Cabin"].copy()
test_data = test_data.drop("Cabin", axis = 1)

In [1048]:
train_data

Pclass                                               Name     Sex   Age  \
0         3                            Braund, Mr. Owen Harris    male  22.0   
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
2         3                             Heikkinen, Miss. Laina  female  26.0   
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
4         3                           Allen, Mr. William Henry    male  35.0   
5         3                                   Moran, Mr. James    male   NaN   
6         1                            McCarthy, Mr. Timothy J    male  54.0   
7         3                     Palsson, Master. Gosta Leonard    male   2.0   
8         3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
9         2                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
10        3                    Sandstrom, Miss. Marguerite Rut  female   4.0   
11        1                           Bonnell, Miss. Elizabeth  female  58.0   
12        3                     Saundercock, Mr. William Henry    male  20.0   
13        3                        Andersson, Mr. Anders Johan    male  39.0   
14        3               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
15        2                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
16        3                               Rice, Master. Eugene    male   2.0   
17        2                       Williams, Mr. Charles Eugene    male   NaN   
18        3  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0   
19        3                            Masselmani, Mrs. Fatima  female   NaN   
20        2                               Fynney, Mr. Joseph J    male  35.0   
21        2                              Beesley, Mr. Lawrence    male  34.0   
22        3                        McGowan, Miss. Anna "Annie"  female  15.0   
23        1                       Sloper, Mr. William Thompson    male  28.0   
24        3                      Palsson, Miss. Torborg Danira  female   8.0   
25        3  Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0   
26        3                            Emir, Mr. Farred Chehab    male   NaN   
27        1                     Fortune, Mr. Charles Alexander    male  19.0   
28        3                      O'Dwyer, Miss. Ellen "Nellie"  female   NaN   
29        3                                Todoroff, Mr. Lalio    male   NaN   
..      ...                                                ...     ...   ...   
861       2                        Giles, Mr. Frederick Edward    male  21.0   
862       1  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  female  48.0   
863       3                  Sage, Miss. Dorothy Edith "Dolly"  female   NaN   
864       2                             Gill, Mr. John William    male  24.0   
865       2                           Bystrom, Mrs. (Karolina)  female  42.0   
866       2                       Duran y More, Miss. Asuncion  female  27.0   
867       1               Roebling, Mr. Washington Augustus II    male  31.0   
868       3                        van Melkebeke, Mr. Philemon    male   NaN   
869       3                    Johnson, Master. Harold Theodor    male   4.0   
870       3                                  Balkic, Mr. Cerin    male  26.0   
871       1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0   
872       1                           Carlsson, Mr. Frans Olof    male  33.0   
873       3                        Vander Cruyssen, Mr. Victor    male  47.0   
874       2              Abelson, Mrs. Samuel (Hannah Wizosky)  female  28.0   
875       3                   Najib, Miss. Adele Kiamie "Jane"  female  15.0   
876       3                      Gustafsson, Mr. Alfred Ossian    male  20.0   
877       3                               Petroff, Mr. Nedelio    male  19.0   
878       3                                 Laleff, Mr. Kristo    male   NaN   
879  

In [1049]:
test_data

Pclass                                               Name     Sex   Age  \
0         3                                   Kelly, Mr. James    male  34.5   
1         3                   Wilkes, Mrs. James (Ellen Needs)  female  47.0   
2         2                          Myles, Mr. Thomas Francis    male  62.0   
3         3                                   Wirz, Mr. Albert    male  27.0   
4         3       Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0   
5         3                         Svensson, Mr. Johan Cervin    male  14.0   
6         3                               Connolly, Miss. Kate  female  30.0   
7         2                       Caldwell, Mr. Albert Francis    male  26.0   
8         3          Abrahim, Mrs. Joseph (Sophie Halaut Easu)  female  18.0   
9         3                            Davies, Mr. John Samuel    male  21.0   
10        3                                   Ilieff, Mr. Ylio    male   NaN   
11        1                         Jones, Mr. Charles Cresson    male  46.0   
12        1      Snyder, Mrs. John Pillsbury (Nelle Stevenson)  female  23.0   
13        2                               Howard, Mr. Benjamin    male  63.0   
14        1  Chaffee, Mrs. Herbert Fuller (Carrie Constance...  female  47.0   
15        2      del Carlo, Mrs. Sebastiano (Argenia Genovesi)  female  24.0   
16        2                                  Keane, Mr. Daniel    male  35.0   
17        3                                  Assaf, Mr. Gerios    male  21.0   
18        3                       Ilmakangas, Miss. Ida Livija  female  27.0   
19        3              Assaf Khalil, Mrs. Mariana (Miriam")"  female  45.0   
20        1                             Rothschild, Mr. Martin    male  55.0   
21        3                          Olsen, Master. Artur Karl    male   9.0   
22        1               Flegenheim, Mrs. Alfred (Antoinette)  female   NaN   
23        1                    Williams, Mr. Richard Norris II    male  21.0   
24        1    Ryerson, Mrs. Arthur Larned (Emily Maria Borie)  female  48.0   
25        3                            Robins, Mr. Alexander A    male  50.0   
26        1                       Ostby, Miss. Helene Ragnhild  female  22.0   
27        3                                  Daher, Mr. Shedid    male  22.5   
28        1                            Brady, Mr. John Bertram    male  41.0   
29        3                                  Samaan, Mr. Elias    male   NaN   
..      ...                                                ...     ...   ...   
388       3                               Canavan, Mr. Patrick    male  21.0   
389       3                        Palsson, Master. Paul Folke    male   6.0   
390       1                         Payne, Mr. Vivian Ponsonby    male  23.0   
391       1     Lines, Mrs. Ernest H (Elizabeth Lindsey James)  female  51.0   
392       3                      Abbott, Master. Eugene Joseph    male  13.0   
393       2                               Gilbert, Mr. William    male  47.0   
394       3                           Kink-Heilmann, Mr. Anton    male  29.0   
395       1     Smith, Mrs. Lucien Philip (Mary Eloise Hughes)  female  18.0   
396       3                               Colbert, Mr. Patrick    male  24.0   
397       1  Frolicher-Stehli, Mrs. Maxmillian (Margaretha ...  female  48.0   
398       3                     Larsson-Rondberg, Mr. Edvard A    male  22.0   
399       3                           Conlon, Mr. Thomas Henry    male  31.0   
400       1                            Bonnell, Miss. Caroline  female  30.0   
401       2                                    Gale, Mr. Harry    male  38.0   
402       1                     Gibson, Miss. Dorothy Winifred  female  22.0   
403       1                             Carrau, Mr. Jose Pedro    male  17.0   
404       1                       Frauenthal, Mr. Isaac Gerald    male  43.0   
405       2       Nourney, Mr. Alfred (Baron von Drachstedt")"    male  20.0   
406  

# Include new columns "Fare/Age" and "Fare/Pclass" and "Age/Pclass" and "RelativesOnBoard" and "AgeBucket" and "TicketBucket"

- Train Data

In [1050]:
train_data["FarePerAge"] = train_data["Fare"] / train_data["Age"]

- Test Data

In [1051]:
test_data["FarePerAge"] = test_data["Fare"] / test_data["Age"]

- Train Data

In [1052]:
train_data["FarePerPclass"] = train_data["Fare"] / train_data["Pclass"]

- Test Data

In [1053]:
test_data["FarePerPclass"] = test_data["Fare"] / test_data["Pclass"]

- Train Data

In [1054]:
train_data["AgePerClass"] = train_data["FarePerPclass"] / train_data["FarePerAge"]

- Test Data

In [1055]:
test_data["AgePerClass"] = test_data["FarePerPclass"] / test_data["FarePerAge"]

- Train Data

In [1056]:
train_data["RelativesOnboard"] = train_data["SibSp"] + train_data["Parch"]

- Test Data

In [1057]:
test_data["RelativesOnboard"] = test_data["SibSp"] + test_data["Parch"]

- Train Data

In [1058]:
train_data["AgeBucket"] = train_data["Age"] // 10 * 10

- Test Data

In [1059]:
test_data["AgeBucket"] = test_data["Age"] // 10 * 10

- Train Data

In [1060]:
train_data["TicketPriceBucket"] = train_data["Fare"] // 5 * 5

- Test Data

In [1061]:
test_data["TicketPriceBucket"] = test_data["Fare"] // 5 * 5

## Turn Pclass into binary values

- Train data

In [1062]:
train_data = pd.concat([train_data, pd.get_dummies(train_data["Pclass"], prefix = "Pclass")], axis = 1)
train_data = train_data.drop("Pclass", axis = 1)

"""def dummy_data(data, columns):
    for column in columns:
        data = pd.concat([data, pd.get_dummies(data[column], prefix=column)], axis=1)
        data = data.drop(column, axis=1)
    return data


dummy_columns = ["Pclass"]
train_data=dummy_data(train_data, dummy_columns)
test_data=dummy_data(test_data, dummy_columns)"""

'def dummy_data(data, columns):\n    for column in columns:\n        data = pd.concat([data, pd.get_dummies(data[column], prefix=column)], axis=1)\n        data = data.drop(column, axis=1)\n    return data\n\n\ndummy_columns = ["Pclass"]\ntrain_data=dummy_data(train_data, dummy_columns)\ntest_data=dummy_data(test_data, dummy_columns)'

- Test data

In [1063]:
test_data = pd.concat([test_data, pd.get_dummies(test_data["Pclass"], prefix = "Pclass")], axis = 1)
test_data = test_data.drop("Pclass", axis = 1)

## Let's binarize the gender

- Train data

In [1064]:
train_data = pd.concat([train_data, pd.get_dummies(train_data["Sex"], prefix = "Sex")], axis = 1)
train_data = train_data.drop("Sex", axis = 1)

- Test data

In [1065]:
test_data = pd.concat([test_data, pd.get_dummies(test_data["Sex"], prefix = "Sex")], axis = 1)
test_data = test_data.drop("Sex", axis = 1)

In [1066]:
train_data

Name   Age  SibSp  Parch  \
0                              Braund, Mr. Owen Harris  22.0      1      0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1      0   
2                               Heikkinen, Miss. Laina  26.0      0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1      0   
4                             Allen, Mr. William Henry  35.0      0      0   
5                                     Moran, Mr. James   NaN      0      0   
6                              McCarthy, Mr. Timothy J  54.0      0      0   
7                       Palsson, Master. Gosta Leonard   2.0      3      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  27.0      0      2   
9                  Nasser, Mrs. Nicholas (Adele Achem)  14.0      1      0   
10                     Sandstrom, Miss. Marguerite Rut   4.0      1      1   
11                            Bonnell, Miss. Elizabeth  58.0      0      0   
12                      Saundercock, Mr. William Henry  20.0      0      0   
13                         Andersson, Mr. Anders Johan  39.0      1      5   
14                Vestrom, Miss. Hulda Amanda Adolfina  14.0      0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   55.0      0      0   
16                                Rice, Master. Eugene   2.0      4      1   
17                        Williams, Mr. Charles Eugene   NaN      0      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  31.0      1      0   
19                             Masselmani, Mrs. Fatima   NaN      0      0   
20                                Fynney, Mr. Joseph J  35.0      0      0   
21                               Beesley, Mr. Lawrence  34.0      0      0   
22                         McGowan, Miss. Anna "Annie"  15.0      0      0   
23                        Sloper, Mr. William Thompson  28.0      0      0   
24                       Palsson, Miss. Torborg Danira   8.0      3      1   
25   Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  38.0      1      5   
26                             Emir, Mr. Farred Chehab   NaN      0      0   
27                      Fortune, Mr. Charles Alexander  19.0      3      2   
28                       O'Dwyer, Miss. Ellen "Nellie"   NaN      0      0   
29                                 Todoroff, Mr. Lalio   NaN      0      0   
..                                                 ...   ...    ...    ...   
861                        Giles, Mr. Frederick Edward  21.0      1      0   
862  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  48.0      0      0   
863                  Sage, Miss. Dorothy Edith "Dolly"   NaN      8      2   
864                             Gill, Mr. John William  24.0      0      0   
865                           Bystrom, Mrs. (Karolina)  42.0      0      0   
866                       Duran y More, Miss. Asuncion  27.0      1      0   
867               Roebling, Mr. Washington Augustus II  31.0      0      0   
868                        van Melkebeke, Mr. Philemon   NaN      0      0   
869                    Johnson, Master. Harold Theodor   4.0      1      1   
870                                  Balkic, Mr. Cerin  26.0      0      0   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  47.0      1      1   
872                           Carlsson, Mr. Frans Olof  33.0      0      0   
873                        Vander Cruyssen, Mr. Victor  47.0      0      0   
874              Abelson, Mrs. Samuel (Hannah Wizosky)  28.0      1      0   
875                   Najib, Miss. Adele Kiamie "Jane"  15.0      0      0   
876                      Gustafsson, Mr. Alfred Ossian  20.0      0      0   
877                               Petroff, Mr. Nedelio  19.0      0      0   
878                                 Laleff, Mr. Kristo   NaN      0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  56.0      0      1   
880       Shelley, Mrs. William (Imanita Parrish Hall)  25.0      0     

In [1067]:
train_data.shape

(891, 17)

In [1068]:
test_data

Name   Age  SibSp  Parch  \
0                                     Kelly, Mr. James  34.5      0      0   
1                     Wilkes, Mrs. James (Ellen Needs)  47.0      1      0   
2                            Myles, Mr. Thomas Francis  62.0      0      0   
3                                     Wirz, Mr. Albert  27.0      0      0   
4         Hirvonen, Mrs. Alexander (Helga E Lindqvist)  22.0      1      1   
5                           Svensson, Mr. Johan Cervin  14.0      0      0   
6                                 Connolly, Miss. Kate  30.0      0      0   
7                         Caldwell, Mr. Albert Francis  26.0      1      1   
8            Abrahim, Mrs. Joseph (Sophie Halaut Easu)  18.0      0      0   
9                              Davies, Mr. John Samuel  21.0      2      0   
10                                    Ilieff, Mr. Ylio   NaN      0      0   
11                          Jones, Mr. Charles Cresson  46.0      0      0   
12       Snyder, Mrs. John Pillsbury (Nelle Stevenson)  23.0      1      0   
13                                Howard, Mr. Benjamin  63.0      1      0   
14   Chaffee, Mrs. Herbert Fuller (Carrie Constance...  47.0      1      0   
15       del Carlo, Mrs. Sebastiano (Argenia Genovesi)  24.0      1      0   
16                                   Keane, Mr. Daniel  35.0      0      0   
17                                   Assaf, Mr. Gerios  21.0      0      0   
18                        Ilmakangas, Miss. Ida Livija  27.0      1      0   
19               Assaf Khalil, Mrs. Mariana (Miriam")"  45.0      0      0   
20                              Rothschild, Mr. Martin  55.0      1      0   
21                           Olsen, Master. Artur Karl   9.0      0      1   
22                Flegenheim, Mrs. Alfred (Antoinette)   NaN      0      0   
23                     Williams, Mr. Richard Norris II  21.0      0      1   
24     Ryerson, Mrs. Arthur Larned (Emily Maria Borie)  48.0      1      3   
25                             Robins, Mr. Alexander A  50.0      1      0   
26                        Ostby, Miss. Helene Ragnhild  22.0      0      1   
27                                   Daher, Mr. Shedid  22.5      0      0   
28                             Brady, Mr. John Bertram  41.0      0      0   
29                                   Samaan, Mr. Elias   NaN      2      0   
..                                                 ...   ...    ...    ...   
388                               Canavan, Mr. Patrick  21.0      0      0   
389                        Palsson, Master. Paul Folke   6.0      3      1   
390                         Payne, Mr. Vivian Ponsonby  23.0      0      0   
391     Lines, Mrs. Ernest H (Elizabeth Lindsey James)  51.0      0      1   
392                      Abbott, Master. Eugene Joseph  13.0      0      2   
393                               Gilbert, Mr. William  47.0      0      0   
394                           Kink-Heilmann, Mr. Anton  29.0      3      1   
395     Smith, Mrs. Lucien Philip (Mary Eloise Hughes)  18.0      1      0   
396                               Colbert, Mr. Patrick  24.0      0      0   
397  Frolicher-Stehli, Mrs. Maxmillian (Margaretha ...  48.0      1      1   
398                     Larsson-Rondberg, Mr. Edvard A  22.0      0      0   
399                           Conlon, Mr. Thomas Henry  31.0      0      0   
400                            Bonnell, Miss. Caroline  30.0      0      0   
401                                    Gale, Mr. Harry  38.0      1      0   
402                     Gibson, Miss. Dorothy Winifred  22.0      0      1   
403                             Carrau, Mr. Jose Pedro  17.0      0      0   
404                       Frauenthal, Mr. Isaac Gerald  43.0      1      0   
405       Nourney, Mr. Alfred (Baron von Drachstedt")"  20.0      0      0   
406                          Ware, Mr. William Jeffery  23.0      1      0   
407                         Widener, Mr. George Dunton  50.0      1     

## Name engineering

- Train Data

In [1069]:
NameSplit = train_data.Name.str.split('[,.]')

In [1070]:
NameSplit.head()

0                          [Braund,  Mr,  Owen Harris]
1    [Cumings,  Mrs,  John Bradley (Florence Briggs...
2                           [Heikkinen,  Miss,  Laina]
3     [Futrelle,  Mrs,  Jacques Heath (Lily May Peel)]
4                         [Allen,  Mr,  William Henry]
Name: Name, dtype: object

In [1071]:
titles = [str.strip(name[1]) for name in NameSplit.values]
titles[:10]

['Mr', 'Mrs', 'Miss', 'Mrs', 'Mr', 'Mr', 'Mr', 'Master', 'Mrs', 'Mrs']

In [1072]:
# New feature
train_data["Title"] = titles

In [1073]:
train_data.Title.unique()

array(['Mr', 'Mrs', 'Miss', 'Master', 'Don', 'Rev', 'Dr', 'Mme', 'Ms',
       'Major', 'Lady', 'Sir', 'Mlle', 'Col', 'Capt', 'the Countess',
       'Jonkheer'], dtype=object)

In [1074]:
# redundancy: combine Mademoiselle and Madame into a single type
train_data.Title.values[train_data.Title.isin(['Mme', 'Mlle'])] = 'Mlle'

In [1075]:
# keep reducing the number of factor levels
train_data.Title.values[train_data.Title.isin(['Capt', 'Don', 'Major', 'Sir'])] = 'Sir'
train_data.Title.values[train_data.Title.isin(['Dona', 'Lady', 'the Countess', 'Jonkheer'])] = 'Lady'

In [1076]:
train_data.head()

Name   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1      0   
2                             Heikkinen, Miss. Laina  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1      0   
4                           Allen, Mr. William Henry  35.0      0      0   

      Fare Embarked  FarePerAge  FarePerPclass  AgePerClass  RelativesOnboard  \
0   7.2500        S    0.329545       2.416667     7.333333                 1   
1  71.2833        C    1.875876      71.283300    38.000000                 1   
2   7.9250        S    0.304808       2.641667     8.666667                 0   
3  53.1000        S    1.517143      53.100000    35.000000                 1   
4   8.0500        S    0.230000       2.683333    11.666667                 0   

   AgeBucket  TicketPriceBucket  Pclass_1  Pclass_2  Pclass_3  Sex_female  \
0       20.0                5.0         0         0         1           0   
1       30.0               70.0         1         0         0           1   
2       20.0                5.0         0         0         1           1   
3       30.0               50.0         1         0         0           1   
4       30.0                5.0         0         0         1           0   

   Sex_male Title  
0         1    Mr  
1         0   Mrs  
2         0  Miss  
3         0   Mrs  
4         1    Mr

- Test Data

In [1077]:
NameSplit = test_data.Name.str.split('[,.]')

In [1078]:
NameSplit.head()

0                                [Kelly,  Mr,  James]
1                [Wilkes,  Mrs,  James (Ellen Needs)]
2                       [Myles,  Mr,  Thomas Francis]
3                                [Wirz,  Mr,  Albert]
4    [Hirvonen,  Mrs,  Alexander (Helga E Lindqvist)]
Name: Name, dtype: object

In [1079]:
titles = [str.strip(name[1]) for name in NameSplit.values]
titles[:10]

['Mr', 'Mrs', 'Mr', 'Mr', 'Mrs', 'Mr', 'Miss', 'Mr', 'Mrs', 'Mr']

In [1080]:
# New feature
test_data["Title"] = titles

In [1081]:
test_data.Title.unique()

array(['Mr', 'Mrs', 'Miss', 'Master', 'Ms', 'Col', 'Rev', 'Dr', 'Dona'],
      dtype=object)

In [1082]:
# redundancy: combine Mademoiselle and Madame into a single type
test_data.Title.values[test_data.Title.isin(['Mme', 'Mlle'])] = 'Mlle'

In [1083]:
# keep reducing the number of factor levels
test_data.Title.values[test_data.Title.isin(['Capt', 'Don', 'Major', 'Sir'])] = 'Sir'
test_data.Title.values[test_data.Title.isin(['Dona', 'Lady', 'the Countess', 'Jonkheer'])] = 'Lady'

In [1084]:
test_data.head()

Name   Age  SibSp  Parch     Fare  \
0                              Kelly, Mr. James  34.5      0      0   7.8292   
1              Wilkes, Mrs. James (Ellen Needs)  47.0      1      0   7.0000   
2                     Myles, Mr. Thomas Francis  62.0      0      0   9.6875   
3                              Wirz, Mr. Albert  27.0      0      0   8.6625   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  22.0      1      1  12.2875   

  Embarked  FarePerAge  FarePerPclass  AgePerClass  RelativesOnboard  \
0        Q    0.226933       2.609733    11.500000                 0   
1        S    0.148936       2.333333    15.666667                 1   
2        Q    0.156250       4.843750    31.000000                 0   
3        S    0.320833       2.887500     9.000000                 0   
4        S    0.558523       4.095833     7.333333                 2   

   AgeBucket  TicketPriceBucket  Pclass_1  Pclass_2  Pclass_3  Sex_female  \
0       30.0                5.0         0         0         1           0   
1       40.0                5.0         0         0         1           1   
2       60.0                5.0         0         1         0           0   
3       20.0                5.0         0         0         1           0   
4       20.0               10.0         0         0         1           1   

   Sex_male Title  
0         1    Mr  
1         0   Mrs  
2         1    Mr  
3         1    Mr  
4         0   Mrs

## Drop "Name"

- Train data

In [1085]:
name_train = train_data["Name"].copy()
train_data = train_data.drop("Name", axis = 1)

- Test Data

In [1086]:
name_test = test_data["Name"].copy()
test_data = test_data.drop("Name", axis = 1)

In [1087]:
train_data

Age  SibSp  Parch      Fare Embarked  FarePerAge  FarePerPclass  \
0    22.0      1      0    7.2500        S    0.329545       2.416667   
1    38.0      1      0   71.2833        C    1.875876      71.283300   
2    26.0      0      0    7.9250        S    0.304808       2.641667   
3    35.0      1      0   53.1000        S    1.517143      53.100000   
4    35.0      0      0    8.0500        S    0.230000       2.683333   
5     NaN      0      0    8.4583        Q         NaN       2.819433   
6    54.0      0      0   51.8625        S    0.960417      51.862500   
7     2.0      3      1   21.0750        S   10.537500       7.025000   
8    27.0      0      2   11.1333        S    0.412344       3.711100   
9    14.0      1      0   30.0708        C    2.147914      15.035400   
10    4.0      1      1   16.7000        S    4.175000       5.566667   
11   58.0      0      0   26.5500        S    0.457759      26.550000   
12   20.0      0      0    8.0500        S    0.402500       2.683333   
13   39.0      1      5   31.2750        S    0.801923      10.425000   
14   14.0      0      0    7.8542        S    0.561014       2.618067   
15   55.0      0      0   16.0000        S    0.290909       8.000000   
16    2.0      4      1   29.1250        Q   14.562500       9.708333   
17    NaN      0      0   13.0000        S         NaN       6.500000   
18   31.0      1      0   18.0000        S    0.580645       6.000000   
19    NaN      0      0    7.2250        C         NaN       2.408333   
20   35.0      0      0   26.0000        S    0.742857      13.000000   
21   34.0      0      0   13.0000        S    0.382353       6.500000   
22   15.0      0      0    8.0292        Q    0.535280       2.676400   
23   28.0      0      0   35.5000        S    1.267857      35.500000   
24    8.0      3      1   21.0750        S    2.634375       7.025000   
25   38.0      1      5   31.3875        S    0.825987      10.462500   
26    NaN      0      0    7.2250        C         NaN       2.408333   
27   19.0      3      2  263.0000        S   13.842105     263.000000   
28    NaN      0      0    7.8792        Q         NaN       2.626400   
29    NaN      0      0    7.8958        S         NaN       2.631933   
..    ...    ...    ...       ...      ...         ...            ...   
861  21.0      1      0   11.5000        S    0.547619       5.750000   
862  48.0      0      0   25.9292        S    0.540192      25.929200   
863   NaN      8      2   69.5500        S         NaN      23.183333   
864  24.0      0      0   13.0000        S    0.541667       6.500000   
865  42.0      0      0   13.0000        S    0.309524       6.500000   
866  27.0      1      0   13.8583        C    0.513270       6.929150   
867  31.0      0      0   50.4958        S    1.628897      50.495800   
868   NaN      0      0    9.5000        S         NaN       3.166667   
869   4.0      1      1   11.1333        S    2.783325       3.711100   
870  26.0      0      0    7.8958        S    0.303685       2.631933   
871  47.0      1      1   52.5542        S    1.118174      52.554200   
872  33.0      0      0    5.0000        S    0.151515       5.000000   
873  47.0      0      0    9.0000        S    0.191489       3.000000   
874  28.0      1      0   24.0000        C    0.857143      12.000000   
875  15.0      0      0    7.2250        C    0.481667       2.408333   
876  20.0      0      0    9.8458        S    0.492290       3.281933   
877  19.0      0      0    7.8958        S    0.415568       2.631933   
878   NaN      0      0    7.8958        S         NaN       2.631933   
879  56.0      0      1   83.1583        C    1.484970      83.158300   
880  25.0      0      1   26.0000        S    1.040000      13.000000   
881  33.0      0      0    7.8958        S    0.239267       2.631933   
882  22.0      0      0   10.5167        S    0.478032       3.505567   
883  28.0      0      0   10.5000        S    0.375000       5.2

In [1088]:
test_data.head()

Age  SibSp  Parch     Fare Embarked  FarePerAge  FarePerPclass  \
0  34.5      0      0   7.8292        Q    0.226933       2.609733   
1  47.0      1      0   7.0000        S    0.148936       2.333333   
2  62.0      0      0   9.6875        Q    0.156250       4.843750   
3  27.0      0      0   8.6625        S    0.320833       2.887500   
4  22.0      1      1  12.2875        S    0.558523       4.095833   

   AgePerClass  RelativesOnboard  AgeBucket  TicketPriceBucket  Pclass_1  \
0    11.500000                 0       30.0                5.0         0   
1    15.666667                 1       40.0                5.0         0   
2    31.000000                 0       60.0                5.0         0   
3     9.000000                 0       20.0                5.0         0   
4     7.333333                 2       20.0               10.0         0   

   Pclass_2  Pclass_3  Sex_female  Sex_male Title  
0         0         1           0         1    Mr  
1         0         1           1         0   Mrs  
2         1         0           0         1    Mr  
3         0         1           0         1    Mr  
4         0         1           1         0   Mrs

# Pipeline

- 1-hot Enconder
- Normalizer
- Imputer (to deal with the NaN values)
- Union

In [1089]:
# Background magic

# The CategoricalEncoder class will allow us to convert categorical attributes to one-hot vectors.

from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.utils import check_array
from sklearn.preprocessing import LabelEncoder
from scipy import sparse

class CategoricalEncoder(BaseEstimator, TransformerMixin):
    """Encode categorical features as a numeric array.
    The input to this transformer should be a matrix of integers or strings,
    denoting the values taken on by categorical (discrete) features.
    The features can be encoded using a one-hot aka one-of-K scheme
    (``encoding='onehot'``, the default) or converted to ordinal integers
    (``encoding='ordinal'``).
    This encoding is needed for feeding categorical data to many scikit-learn
    estimators, notably linear models and SVMs with the standard kernels.
    Read more in the :ref:`User Guide <preprocessing_categorical_features>`.
    Parameters
    ----------
    encoding : str, 'onehot', 'onehot-dense' or 'ordinal'
        The type of encoding to use (default is 'onehot'):
        - 'onehot': encode the features using a one-hot aka one-of-K scheme
          (or also called 'dummy' encoding). This creates a binary column for
          each category and returns a sparse matrix.
        - 'onehot-dense': the same as 'onehot' but returns a dense array
          instead of a sparse matrix.
        - 'ordinal': encode the features as ordinal integers. This results in
          a single column of integers (0 to n_categories - 1) per feature.
    categories : 'auto' or a list of lists/arrays of values.
        Categories (unique values) per feature:
        - 'auto' : Determine categories automatically from the training data.
        - list : ``categories[i]`` holds the categories expected in the ith
          column. The passed categories are sorted before encoding the data
          (used categories can be found in the ``categories_`` attribute).
    dtype : number type, default np.float64
        Desired dtype of output.
    handle_unknown : 'error' (default) or 'ignore'
        Whether to raise an error or ignore if a unknown categorical feature is
        present during transform (default is to raise). When this is parameter
        is set to 'ignore' and an unknown category is encountered during
        transform, the resulting one-hot encoded columns for this feature
        will be all zeros.
        Ignoring unknown categories is not supported for
        ``encoding='ordinal'``.
    Attributes
    ----------
    categories_ : list of arrays
        The categories of each feature determined during fitting. When
        categories were specified manually, this holds the sorted categories
        (in order corresponding with output of `transform`).
    Examples
    --------
    Given a dataset with three features and two samples, we let the encoder
    find the maximum value per feature and transform the data to a binary
    one-hot encoding.
    >>> from sklearn.preprocessing import CategoricalEncoder
    >>> enc = CategoricalEncoder(handle_unknown='ignore')
    >>> enc.fit([[0, 0, 3], [1, 1, 0], [0, 2, 1], [1, 0, 2]])
    ... # doctest: +ELLIPSIS
    CategoricalEncoder(categories='auto', dtype=<... 'numpy.float64'>,
              encoding='onehot', handle_unknown='ignore')
    >>> enc.transform([[0, 1, 1], [1, 0, 4]]).toarray()
    array([[ 1.,  0.,  0.,  1.,  0.,  0.,  1.,  0.,  0.],
           [ 0.,  1.,  1.,  0.,  0.,  0.,  0.,  0.,  0.]])
    See also
    --------
    sklearn.preprocessing.OneHotEncoder : performs a one-hot encoding of
      integer ordinal features. The ``OneHotEncoder assumes`` that input
      features take on values in the range ``[0, max(feature)]`` instead of
      using the unique values.
    sklearn.feature_extraction.DictVectorizer : performs a one-hot encoding of
      dictionary items (also handles string-valued features).
    sklearn.feature_extraction.FeatureHasher : performs an approximate one-hot
      encoding of dictionary items or strings.
    """

    def __init__(self, encoding='onehot', categories='auto', dtype=np.float64,
                 handle_unknown='error'):
        self.encoding = encoding
        self.categories = categories
        self.dtype = dtype
        self.handle_unknown = handle_unknown

    def fit(self, X, y=None):
        """Fit the CategoricalEncoder to X.
        Parameters
        ----------
        X : array-like, shape [n_samples, n_feature]
            The data to determine the categories of each feature.
        Returns
        -------
        self
        """

        if self.encoding not in ['onehot', 'onehot-dense', 'ordinal']:
            template = ("encoding should be either 'onehot', 'onehot-dense' "
                        "or 'ordinal', got %s")
            raise ValueError(template % self.handle_unknown)

        if self.handle_unknown not in ['error', 'ignore']:
            template = ("handle_unknown should be either 'error' or "
                        "'ignore', got %s")
            raise ValueError(template % self.handle_unknown)

        if self.encoding == 'ordinal' and self.handle_unknown == 'ignore':
            raise ValueError("handle_unknown='ignore' is not supported for"
                             " encoding='ordinal'")

        X = check_array(X, dtype=np.object, accept_sparse='csc', copy=True)
        n_samples, n_features = X.shape

        self._label_encoders_ = [LabelEncoder() for _ in range(n_features)]

        for i in range(n_features):
            le = self._label_encoders_[i]
            Xi = X[:, i]
            if self.categories == 'auto':
                le.fit(Xi)
            else:
                valid_mask = np.in1d(Xi, self.categories[i])
                if not np.all(valid_mask):
                    if self.handle_unknown == 'error':
                        diff = np.unique(Xi[~valid_mask])
                        msg = ("Found unknown categories {0} in column {1}"
                               " during fit".format(diff, i))
                        raise ValueError(msg)
                le.classes_ = np.array(np.sort(self.categories[i]))

        self.categories_ = [le.classes_ for le in self._label_encoders_]

        return self

    def transform(self, X):
        """Transform X using one-hot encoding.
        Parameters
        ----------
        X : array-like, shape [n_samples, n_features]
            The data to encode.
        Returns
        -------
        X_out : sparse matrix or a 2-d array
            Transformed input.
        """
        X = check_array(X, accept_sparse='csc', dtype=np.object, copy=True)
        n_samples, n_features = X.shape
        X_int = np.zeros_like(X, dtype=np.int)
        X_mask = np.ones_like(X, dtype=np.bool)

        for i in range(n_features):
            valid_mask = np.in1d(X[:, i], self.categories_[i])

            if not np.all(valid_mask):
                if self.handle_unknown == 'error':
                    diff = np.unique(X[~valid_mask, i])
                    msg = ("Found unknown categories {0} in column {1}"
                           " during transform".format(diff, i))
                    raise ValueError(msg)
                else:
                    # Set the problematic rows to an acceptable value and
                    # continue `The rows are marked `X_mask` and will be
                    # removed later.
                    X_mask[:, i] = valid_mask
                    X[:, i][~valid_mask] = self.categories_[i][0]
            X_int[:, i] = self._label_encoders_[i].transform(X[:, i])

        if self.encoding == 'ordinal':
            return X_int.astype(self.dtype, copy=False)

        mask = X_mask.ravel()
        n_values = [cats.shape[0] for cats in self.categories_]
        n_values = np.array([0] + n_values)
        indices = np.cumsum(n_values)

        column_indices = (X_int + indices[:-1]).ravel()[mask]
        row_indices = np.repeat(np.arange(n_samples, dtype=np.int32),
                                n_features)[mask]
        data = np.ones(n_samples * n_features)[mask]

        out = sparse.csc_matrix((data, (row_indices, column_indices)),
                                shape=(n_samples, indices[-1]),
                                dtype=self.dtype).tocsr()
        if self.encoding == 'onehot-dense':
            return out.toarray()
        else:
            return out


In [1090]:
from sklearn.base import BaseEstimator, TransformerMixin

# A class to select numerical or categorical columns 
# since Scikit-Learn doesn't handle DataFrames yet
class DataFrameSelector(BaseEstimator, TransformerMixin):
    def __init__(self, attribute_names):
        self.attribute_names = attribute_names
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        return X[self.attribute_names]

Let's create an imputer for the categorical values as the normal one would not work

In [1091]:
# Inspired from stackoverflow.com/questions/25239958
class MostFrequentImputer(BaseEstimator, TransformerMixin):
    def fit(self, X, y=None):
        self.most_frequent = pd.Series([X[c].value_counts().index[0] for c in X],
                                       index=X.columns)
        return self
    def transform(self, X, y=None):
        return X.fillna(self.most_frequent)

In [1092]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 17 columns):
Age                  714 non-null float64
SibSp                891 non-null int64
Parch                891 non-null int64
Fare                 891 non-null float64
Embarked             889 non-null object
FarePerAge           714 non-null float64
FarePerPclass        891 non-null float64
AgePerClass          707 non-null float64
RelativesOnboard     891 non-null int64
AgeBucket            714 non-null float64
TicketPriceBucket    891 non-null float64
Pclass_1             891 non-null uint8
Pclass_2             891 non-null uint8
Pclass_3             891 non-null uint8
Sex_female           891 non-null uint8
Sex_male             891 non-null uint8
Title                891 non-null object
dtypes: float64(7), int64(3), object(2), uint8(5)
memory usage: 88.0+ KB


In [1093]:
#############
# PIPELINE #
############

from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MinMaxScaler
from sklearn.utils import check_array
from sklearn.preprocessing import LabelEncoder
from scipy import sparse
from sklearn.pipeline import FeatureUnion
from sklearn.preprocessing import Imputer

# Let's create an imputer to deal with the missing values

imputer = Imputer(strategy = "median")

# Let's take care of the numerical data

num_pipeline = Pipeline([
        ("select_numeric", DataFrameSelector(["Age", "SibSp", "Parch", "Fare", "FarePerPclass", "FarePerAge",
                                             "Pclass_1", "Pclass_2", "Pclass_3", "Sex_female", "Sex_female",
                                             "AgePerClass", "RelativesOnboard", "AgeBucket", "TicketPriceBucket"])),
        ("imputer", Imputer(strategy="median")),
        ('std_scaler', MinMaxScaler()),
    ])

# Let's take care of the categorical data

#cat_pipeline = Pipeline([
#        ("select_cat", DataFrameSelector(["Pclass", "Sex", "Embarked"])),
#        ("imputer", MostFrequentImputer()),
#        ("cat_encoder", CategoricalEncoder(encoding='onehot-dense')),
#    ])

cat_pipeline = Pipeline([
        ("select_cat", DataFrameSelector(["Embarked", "Title"])),
        ("imputer", MostFrequentImputer()),
        ("cat_encoder", CategoricalEncoder(encoding='onehot-dense')),
    ])

# Let's join the numerical and categorical data using the union

from sklearn.pipeline import FeatureUnion
preprocess_pipeline = FeatureUnion(transformer_list=[
        ("num_pipeline", num_pipeline),
        ("cat_pipeline", cat_pipeline),
    ])

ACTIVATE THE PIPELINE!

In [1094]:
X_train = preprocess_pipeline.fit_transform(train_data)
X_train

array([[0.27117366, 0.125     , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.4722292 , 0.125     , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.32143755, 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.34656949, 0.125     , 0.33333333, ..., 0.        , 0.        ,
        0.        ],
       [0.32143755, 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.39683338, 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [1095]:
X_train.shape

(891, 29)

In [1096]:
X_test = preprocess_pipeline.transform(test_data)

In [1097]:
X_test.shape

(418, 29)

# Classification Time

In [1098]:
from sklearn.svm import SVC # DONE
from scipy.stats import reciprocal, uniform
from sklearn.ensemble import RandomForestClassifier # DONE
from sklearn.ensemble import AdaBoostClassifier # DONE
from sklearn.ensemble import ExtraTreesClassifier # DONE
from sklearn.ensemble import GradientBoostingClassifier # DONE
from sklearn.linear_model import LogisticRegression # DONE
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score

In [1099]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_score, recall_score
import itertools

## Support Vector Classifier

In [1100]:
svc_clf = SVC(random_state = 42)

#kernel = 
#gamma =
#C =

#param_distributions = {'C':[0.1, 1, 10, 100],'gamma':[1,0.1,0.001,0.0001]}
#gird_search_cv_svc = GridSearchCV(svc_clf, param_distributions, cv = 4, scoring='roc_auc', refit = True, verbose = 2,
#                                 n_jobs = -1)
#gird_search_cv_svc.fit(X_train, train_labels)

param_distributions = {"gamma": reciprocal(0.00001, 1), "C": uniform(1000, 1000000)}

rnd_search_cv_svc = RandomizedSearchCV(svc_clf, param_distributions, n_iter=10, verbose=2, n_jobs = 1)
rnd_search_cv_svc.fit(X_train, train_labels)

Fitting 3 folds for each of 10 candidates, totalling 30 fits
[CV] C=693738.1321058419, gamma=0.00775819046816097 ..................
[CV] ... C=693738.1321058419, gamma=0.00775819046816097, total=   1.1s
[CV] C=693738.1321058419, gamma=0.00775819046816097 ..................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    1.1s remaining:    0.0s


[CV] ... C=693738.1321058419, gamma=0.00775819046816097, total=   0.5s
[CV] C=693738.1321058419, gamma=0.00775819046816097 ..................
[CV] ... C=693738.1321058419, gamma=0.00775819046816097, total=   0.4s
[CV] C=696391.7176789206, gamma=0.9834167223160915 ...................
[CV] .... C=696391.7176789206, gamma=0.9834167223160915, total=   0.6s
[CV] C=696391.7176789206, gamma=0.9834167223160915 ...................
[CV] .... C=696391.7176789206, gamma=0.9834167223160915, total=   1.1s
[CV] C=696391.7176789206, gamma=0.9834167223160915 ...................
[CV] .... C=696391.7176789206, gamma=0.9834167223160915, total=   1.4s
[CV] C=800584.6893999415, gamma=0.0019236127372351391 ................
[CV] . C=800584.6893999415, gamma=0.0019236127372351391, total=   0.6s
[CV] C=800584.6893999415, gamma=0.0019236127372351391 ................
[CV] . C=800584.6893999415, gamma=0.0019236127372351391, total=   1.0s
[CV] C=800584.6893999415, gamma=0.0019236127372351391 ................
[CV] .

[Parallel(n_jobs=1)]: Done  30 out of  30 | elapsed:   14.7s finished


RandomizedSearchCV(cv=None, error_score='raise',
          estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=42, shrinking=True,
  tol=0.001, verbose=False),
          fit_params=None, iid=True, n_iter=10, n_jobs=1,
          param_distributions={'gamma': <scipy.stats._distn_infrastructure.rv_frozen object at 0x7fa68fa73240>, 'C': <scipy.stats._distn_infrastructure.rv_frozen object at 0x7fa68f979a90>},
          pre_dispatch='2*n_jobs', random_state=None, refit=True,
          return_train_score='warn', scoring=None, verbose=2)

In [1101]:
print(rnd_search_cv_svc.best_params_)

print(rnd_search_cv_svc.best_estimator_)

{'C': 800584.6893999415, 'gamma': 0.0019236127372351391}
SVC(C=800584.6893999415, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma=0.0019236127372351391,
  kernel='rbf', max_iter=-1, probability=False, random_state=42,
  shrinking=True, tol=0.001, verbose=False)


In [1102]:
svc_randomized = rnd_search_cv_svc.best_estimator_

y_pred_svc_randomized = svc_randomized.predict(X_train)
accuracy_score(train_labels, y_pred_svc_randomized)

0.8608305274971941

In [1103]:
print(confusion_matrix(train_labels, y_pred_svc_randomized))

print("Precision: ", precision_score(train_labels, y_pred_svc_randomized))

print("Recall: ", recall_score(train_labels, y_pred_svc_randomized))

[[516  33]
 [ 91 251]]
Precision:  0.8838028169014085
Recall:  0.7339181286549707


# Random Forest Classifer

In [1104]:
forest_class = RandomForestClassifier(random_state = 42)

n_estimators = [10, 200]
max_features = [0.1, 0.5]
max_depth = [2, 10, 20] 
oob_score = [True, False]
min_samples_split = [0.1, 0.5]
min_samples_leaf = [0.1, 0.5] 
max_leaf_nodes = [2, 10, 100]

param_grid_forest = {'n_estimators' : n_estimators, 'max_features' : max_features,
                     'max_depth' : max_depth, 'min_samples_split' : min_samples_split,
                    'oob_score' : oob_score, 'min_samples_leaf': min_samples_leaf, 
                     'max_leaf_nodes' : max_leaf_nodes}


rand_search_forest = RandomizedSearchCV(forest_class, param_grid_forest, cv = 4, scoring='roc_auc', refit = True,
                                 n_jobs = -1, verbose=2)

rand_search_forest.fit(X_train, train_labels)

Fitting 4 folds for each of 10 candidates, totalling 40 fits
[CV] oob_score=False, n_estimators=10, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=100, max_features=0.1, max_depth=20 
[CV] oob_score=False, n_estimators=10, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=100, max_features=0.1, max_depth=20 
[CV] oob_score=False, n_estimators=10, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=100, max_features=0.1, max_depth=20 
[CV] oob_score=False, n_estimators=10, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=100, max_features=0.1, max_depth=20 
[CV]  oob_score=False, n_estimators=10, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=100, max_features=0.1, max_depth=20, total=   0.2s
[CV]  oob_score=False, n_estimators=10, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=100, max_features=0.1, max_depth=20, total=   0.1s
[CV] oob_score=False, n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.1, max_leaf

/home/isaac/anaconda3/lib/python3.6/site-packages/sklearn/ensemble/forest.py:453: UserWarning: Some inputs do not have OOB scores. This probably means too few trees were used to compute any reliable oob estimates.
  warn("Some inputs do not have OOB scores. "


[CV]  oob_score=False, n_estimators=200, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.5, max_depth=2, total=   0.5s


/home/isaac/anaconda3/lib/python3.6/site-packages/sklearn/ensemble/forest.py:458: RuntimeWarning: invalid value encountered in true_divide
  predictions[k].sum(axis=1)[:, np.newaxis])


[CV] oob_score=True, n_estimators=10, min_samples_split=0.5, min_samples_leaf=0.1, max_leaf_nodes=10, max_features=0.5, max_depth=10 
[CV]  oob_score=True, n_estimators=10, min_samples_split=0.5, min_samples_leaf=0.1, max_leaf_nodes=10, max_features=0.5, max_depth=10, total=   0.1s
[CV]  oob_score=False, n_estimators=200, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.5, max_depth=2, total=   0.4s
[CV] oob_score=True, n_estimators=10, min_samples_split=0.5, min_samples_leaf=0.1, max_leaf_nodes=10, max_features=0.5, max_depth=10 
[CV] oob_score=True, n_estimators=10, min_samples_split=0.5, min_samples_leaf=0.1, max_leaf_nodes=10, max_features=0.5, max_depth=10 


/home/isaac/anaconda3/lib/python3.6/site-packages/sklearn/ensemble/forest.py:453: UserWarning: Some inputs do not have OOB scores. This probably means too few trees were used to compute any reliable oob estimates.
  warn("Some inputs do not have OOB scores. "
/home/isaac/anaconda3/lib/python3.6/site-packages/sklearn/ensemble/forest.py:458: RuntimeWarning: invalid value encountered in true_divide
  predictions[k].sum(axis=1)[:, np.newaxis])


[CV]  oob_score=False, n_estimators=200, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.5, max_depth=2, total=   0.5s
[CV] oob_score=False, n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.5, max_depth=10 
[CV]  oob_score=True, n_estimators=10, min_samples_split=0.5, min_samples_leaf=0.1, max_leaf_nodes=10, max_features=0.5, max_depth=10, total=   0.0s
[CV] oob_score=False, n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.5, max_depth=10 


/home/isaac/anaconda3/lib/python3.6/site-packages/sklearn/ensemble/forest.py:453: UserWarning: Some inputs do not have OOB scores. This probably means too few trees were used to compute any reliable oob estimates.
  warn("Some inputs do not have OOB scores. "
/home/isaac/anaconda3/lib/python3.6/site-packages/sklearn/ensemble/forest.py:453: UserWarning: Some inputs do not have OOB scores. This probably means too few trees were used to compute any reliable oob estimates.
  warn("Some inputs do not have OOB scores. "
/home/isaac/anaconda3/lib/python3.6/site-packages/sklearn/ensemble/forest.py:458: RuntimeWarning: invalid value encountered in true_divide
  predictions[k].sum(axis=1)[:, np.newaxis])
/home/isaac/anaconda3/lib/python3.6/site-packages/sklearn/ensemble/forest.py:458: RuntimeWarning: invalid value encountered in true_divide
  predictions[k].sum(axis=1)[:, np.newaxis])


[CV]  oob_score=True, n_estimators=10, min_samples_split=0.5, min_samples_leaf=0.1, max_leaf_nodes=10, max_features=0.5, max_depth=10, total=   0.1s
[CV]  oob_score=True, n_estimators=10, min_samples_split=0.5, min_samples_leaf=0.1, max_leaf_nodes=10, max_features=0.5, max_depth=10, total=   0.1s
[CV] oob_score=False, n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.5, max_depth=10 
[CV] oob_score=False, n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.5, max_depth=10 
[CV]  oob_score=False, n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.5, max_depth=10, total=   0.4s
[CV]  oob_score=False, n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.5, max_depth=10, total=   0.4s
[CV] oob_score=True, n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.1, max_leaf_nodes=10, max_features=0.1, max_depth=1

[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:    3.3s
[Parallel(n_jobs=-1)]: Done  40 out of  40 | elapsed:    3.4s finished


RandomizedSearchCV(cv=4, error_score='raise',
          estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=42, verbose=0, warm_start=False),
          fit_params=None, iid=True, n_iter=10, n_jobs=-1,
          param_distributions={'n_estimators': [10, 200], 'max_features': [0.1, 0.5], 'max_depth': [2, 10, 20], 'min_samples_split': [0.1, 0.5], 'oob_score': [True, False], 'min_samples_leaf': [0.1, 0.5], 'max_leaf_nodes': [2, 10, 100]},
          pre_dispatch='2*n_jobs', random_state=None, refit=True,
          return_train_score='warn', scoring='roc_auc', verbose=2)

In [1105]:
print(rand_search_forest.best_params_)

print(rand_search_forest.best_estimator_)

{'oob_score': False, 'n_estimators': 10, 'min_samples_split': 0.1, 'min_samples_leaf': 0.1, 'max_leaf_nodes': 100, 'max_features': 0.1, 'max_depth': 20}
RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=20, max_features=0.1, max_leaf_nodes=100,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=0.1, min_samples_split=0.1,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=42, verbose=0, warm_start=False)


In [1106]:
rand_forest_randomized = rand_search_forest.best_estimator_

y_pred_rand_forest_randomized = rand_forest_randomized.predict(X_train)
accuracy_score(train_labels, y_pred_rand_forest_randomized)

0.7687991021324355

In [1107]:
print(confusion_matrix(train_labels, y_pred_rand_forest_randomized))

print("Precision: ", precision_score(train_labels, y_pred_rand_forest_randomized))

print("Recall: ", recall_score(train_labels, y_pred_rand_forest_randomized))

[[512  37]
 [169 173]]
Precision:  0.8238095238095238
Recall:  0.5058479532163743


# AdaBoost Classifier

In [1108]:
ada_class = AdaBoostClassifier(random_state = 42)

n_estimators = [3, 20, 50, 100, 200]
learning_rate = [0.1, 0.5, 0.9]
algorithm = ['SAMME', 'SAMME.R']

param_grid_ada = {'n_estimators' : n_estimators, 'learning_rate' : learning_rate, 'algorithm' : algorithm}

rand_search_ada = RandomizedSearchCV(ada_class, param_grid_ada, cv = 4, scoring='roc_auc', refit = True, n_jobs = -1, verbose = 2)

rand_search_ada.fit(X_train, train_labels)


Fitting 4 folds for each of 10 candidates, totalling 40 fits
[CV] n_estimators=100, learning_rate=0.1, algorithm=SAMME ............
[CV] n_estimators=100, learning_rate=0.1, algorithm=SAMME ............
[CV] n_estimators=100, learning_rate=0.1, algorithm=SAMME ............
[CV] n_estimators=100, learning_rate=0.1, algorithm=SAMME ............
[CV]  n_estimators=100, learning_rate=0.1, algorithm=SAMME, total=   0.3s
[CV] n_estimators=3, learning_rate=0.9, algorithm=SAMME.R ............
[CV]  n_estimators=3, learning_rate=0.9, algorithm=SAMME.R, total=   0.0s
[CV] n_estimators=3, learning_rate=0.9, algorithm=SAMME.R ............
[CV]  n_estimators=100, learning_rate=0.1, algorithm=SAMME, total=   0.3s
[CV] n_estimators=3, learning_rate=0.9, algorithm=SAMME.R ............
[CV]  n_estimators=3, learning_rate=0.9, algorithm=SAMME.R, total=   0.0s
[CV]  n_estimators=3, learning_rate=0.9, algorithm=SAMME.R, total=   0.0s
[CV] n_estimators=3, learning_rate=0.9, algorithm=SAMME.R ............
[

[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:    2.7s
[Parallel(n_jobs=-1)]: Done  40 out of  40 | elapsed:    2.9s finished


RandomizedSearchCV(cv=4, error_score='raise',
          estimator=AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None,
          learning_rate=1.0, n_estimators=50, random_state=42),
          fit_params=None, iid=True, n_iter=10, n_jobs=-1,
          param_distributions={'n_estimators': [3, 20, 50, 100, 200], 'learning_rate': [0.1, 0.5, 0.9], 'algorithm': ['SAMME', 'SAMME.R']},
          pre_dispatch='2*n_jobs', random_state=None, refit=True,
          return_train_score='warn', scoring='roc_auc', verbose=2)

In [1109]:
print(rand_search_ada.best_params_)

print(rand_search_ada.best_estimator_)

{'n_estimators': 50, 'learning_rate': 0.1, 'algorithm': 'SAMME.R'}
AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None,
          learning_rate=0.1, n_estimators=50, random_state=42)


In [1110]:
ada_randomized = rand_search_ada.best_estimator_

y_pred_ada_randomized = ada_randomized.predict(X_train)
accuracy_score(train_labels, y_pred_ada_randomized)

0.819304152637486

In [1111]:
print(confusion_matrix(train_labels, y_pred_ada_randomized))

print("Precision: ", precision_score(train_labels, y_pred_ada_randomized))

print("Recall: ", recall_score(train_labels, y_pred_ada_randomized))

[[485  64]
 [ 97 245]]
Precision:  0.7928802588996764
Recall:  0.716374269005848


Wow, AdaBoost so far has the best Precision and Recall. Nice, nice.

# Extra Trees Classifier

In [1112]:
extra_tree_class = ExtraTreesClassifier(random_state = 42)

n_estimators = [3, 50, 100, 200]
max_features = [0.1, 0.5]
max_depth = [2, 50, 100]
min_samples_split = [0.1, 0.5]
min_samples_leaf = [0.1, 0.5] # Mhm, this one leads to accuracy of test and train sets being the same.

#param_grid_extra_trees = {'n_estimators' : n_estimators, 'max_features' : max_features,
#                          'min_samples_leaf' : min_samples_leaf}
param_grid_extra_trees = {'n_estimators' : n_estimators, 'max_features' : max_features,
                         'max_depth' : max_depth, 'min_samples_split' : min_samples_split,
                         'min_samples_leaf' : min_samples_leaf}


rand_search_extra_trees = RandomizedSearchCV(extra_tree_class, param_grid_extra_trees, cv = 4, scoring='roc_auc', 
                               refit = True, n_jobs = -1, verbose = 2)

rand_search_extra_trees.fit(X_train, train_labels)

Fitting 4 folds for each of 10 candidates, totalling 40 fits
[CV] n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.1, max_features=0.1, max_depth=2 
[CV] n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.1, max_features=0.1, max_depth=2 
[CV] n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.1, max_features=0.1, max_depth=2 
[CV] n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.1, max_features=0.1, max_depth=2 
[CV]  n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.1, max_features=0.1, max_depth=2, total=   0.0s
[CV] n_estimators=3, min_samples_split=0.5, min_samples_leaf=0.1, max_features=0.1, max_depth=2 
[CV]  n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.1, max_features=0.1, max_depth=2, total=   0.0s
[CV]  n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.1, max_features=0.1, max_depth=2, total=   0.0s
[CV]  n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.1, max_features=0.1, max_depth=2, total=   0.0s
[CV] n

[CV]  n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.1, max_features=0.5, max_depth=2, total=   0.2s
[CV] n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.1, max_features=0.5, max_depth=2 
[CV]  n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.1, max_features=0.5, max_depth=2, total=   0.2s


[Parallel(n_jobs=-1)]: Done  40 out of  40 | elapsed:    3.1s finished


RandomizedSearchCV(cv=4, error_score='raise',
          estimator=ExtraTreesClassifier(bootstrap=False, class_weight=None, criterion='gini',
           max_depth=None, max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
           oob_score=False, random_state=42, verbose=0, warm_start=False),
          fit_params=None, iid=True, n_iter=10, n_jobs=-1,
          param_distributions={'n_estimators': [3, 50, 100, 200], 'max_features': [0.1, 0.5], 'max_depth': [2, 50, 100], 'min_samples_split': [0.1, 0.5], 'min_samples_leaf': [0.1, 0.5]},
          pre_dispatch='2*n_jobs', random_state=None, refit=True,
          return_train_score='warn', scoring='roc_auc', verbose=2)

In [1113]:
print(rand_search_extra_trees.best_params_)

print(rand_search_extra_trees.best_estimator_)

{'n_estimators': 50, 'min_samples_split': 0.1, 'min_samples_leaf': 0.1, 'max_features': 0.5, 'max_depth': 2}
ExtraTreesClassifier(bootstrap=False, class_weight=None, criterion='gini',
           max_depth=2, max_features=0.5, max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=0.1, min_samples_split=0.1,
           min_weight_fraction_leaf=0.0, n_estimators=50, n_jobs=1,
           oob_score=False, random_state=42, verbose=0, warm_start=False)


In [1114]:
extra_trees_randomized = rand_search_extra_trees.best_estimator_

y_pred_extra_trees_randomized = extra_trees_randomized.predict(X_train)
accuracy_score(train_labels, y_pred_extra_trees_randomized)

0.7878787878787878

In [1115]:
print(confusion_matrix(train_labels, y_pred_extra_trees_randomized))

print("Precision: ", precision_score(train_labels, y_pred_extra_trees_randomized))

print("Recall: ", recall_score(train_labels, y_pred_extra_trees_randomized))

[[461  88]
 [101 241]]
Precision:  0.7325227963525835
Recall:  0.7046783625730995


# Gradient Boosting Classifier

In [1116]:
grad_boost_class = GradientBoostingClassifier(random_state = 42)

n_estimators = [3, 200]
learning_rate = [0.1, 0.5]
max_depth = [3, 50, 100]
min_samples_split = [0.1, 0.5]
min_samples_leaf = [0.1, 0.5]
max_features = [0.1, 0.5]
max_leaf_nodes = [2, 50, 100]
                            
param_grid_grad_boost_class = {'n_estimators' : n_estimators, 'learning_rate' : learning_rate,
                              'max_depth' : max_depth, 'min_samples_split' : min_samples_split,
                              'min_samples_leaf' : min_samples_leaf, 'max_features' : max_features,
                              'max_leaf_nodes' : max_leaf_nodes}

rand_search_grad_boost_class = RandomizedSearchCV(grad_boost_class, param_grid_grad_boost_class, cv = 4, scoring='roc_auc', 
                               refit = True, n_jobs = -1, verbose = 2)

rand_search_grad_boost_class.fit(X_train, train_labels)

Fitting 4 folds for each of 10 candidates, totalling 40 fits
[CV] n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.1, max_depth=50, learning_rate=0.1 
[CV] n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.1, max_depth=50, learning_rate=0.1 
[CV] n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.1, max_depth=50, learning_rate=0.1 
[CV] n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.1, max_depth=50, learning_rate=0.1 
[CV]  n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.1, max_depth=50, learning_rate=0.1, total=   0.0s
[CV]  n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0.1, max_depth=50, learning_rate=0.1, total=   0.1s
[CV]  n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=2, max_features=0

[CV] n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=50, max_features=0.5, max_depth=50, learning_rate=0.5 
[CV]  n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=50, max_features=0.5, max_depth=50, learning_rate=0.5, total=   0.0s
[CV]  n_estimators=200, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=100, max_features=0.5, max_depth=100, learning_rate=0.5, total=   0.1s
[CV]  n_estimators=200, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=100, max_features=0.5, max_depth=50, learning_rate=0.5, total=   0.1s
[CV]  n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.5, max_leaf_nodes=50, max_features=0.1, max_depth=50, learning_rate=0.1, total=   0.1s
[CV] n_estimators=3, min_samples_split=0.1, min_samples_leaf=0.5, max_leaf_nodes=50, max_features=0.1, max_depth=50, learning_rate=0.5 
[CV] n_estimators=200, min_samples_split=0.5, min_samples_leaf=0.5, max_leaf_nodes=50, max_features=0.1, max_depth=

[Parallel(n_jobs=-1)]: Done  40 out of  40 | elapsed:    1.4s finished


RandomizedSearchCV(cv=4, error_score='raise',
          estimator=GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.1, loss='deviance', max_depth=3,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=2,
              min_weight_fraction_leaf=0.0, n_estimators=100,
              presort='auto', random_state=42, subsample=1.0, verbose=0,
              warm_start=False),
          fit_params=None, iid=True, n_iter=10, n_jobs=-1,
          param_distributions={'n_estimators': [3, 200], 'learning_rate': [0.1, 0.5], 'max_depth': [3, 50, 100], 'min_samples_split': [0.1, 0.5], 'min_samples_leaf': [0.1, 0.5], 'max_features': [0.1, 0.5], 'max_leaf_nodes': [2, 50, 100]},
          pre_dispatch='2*n_jobs', random_state=None, refit=True,
          return_train_score='warn', scoring='roc_auc', verbose=2)

In [1117]:
print(rand_search_grad_boost_class.best_params_)

print(rand_search_grad_boost_class.best_estimator_)

{'n_estimators': 200, 'min_samples_split': 0.1, 'min_samples_leaf': 0.1, 'max_leaf_nodes': 50, 'max_features': 0.1, 'max_depth': 3, 'learning_rate': 0.1}
GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.1, loss='deviance', max_depth=3,
              max_features=0.1, max_leaf_nodes=50,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=0.1, min_samples_split=0.1,
              min_weight_fraction_leaf=0.0, n_estimators=200,
              presort='auto', random_state=42, subsample=1.0, verbose=0,
              warm_start=False)


In [1118]:
grad_boost_randomized = rand_search_grad_boost_class.best_estimator_

y_pred_grad_boost_randomized = grad_boost_randomized.predict(X_train)
accuracy_score(train_labels, y_pred_extra_trees_randomized)

0.7878787878787878

In [1119]:
print(confusion_matrix(train_labels, y_pred_extra_trees_randomized))

print("Precision: ", precision_score(train_labels, y_pred_extra_trees_randomized))

print("Recall: ", recall_score(train_labels, y_pred_extra_trees_randomized))

[[461  88]
 [101 241]]
Precision:  0.7325227963525835
Recall:  0.7046783625730995


# Logistic Regression

In [1120]:
log_reg = LogisticRegression(random_state = 42)

C = np.array(list(range(1, 11)))/10
                            
param_grid_log_reg = {'C' : C}

rand_search_log_reg = RandomizedSearchCV(log_reg, param_grid_log_reg, cv = 4, scoring='roc_auc', 
                               refit = True, n_jobs = -1, verbose = 2)

rand_search_log_reg.fit(X_train, train_labels)

Fitting 4 folds for each of 10 candidates, totalling 40 fits
[CV] C=0.1 ...........................................................
[CV] C=0.1 ...........................................................
[CV] C=0.1 ...........................................................
[CV] C=0.1 ...........................................................
[CV] ............................................ C=0.1, total=   0.2s
[CV] ............................................ C=0.1, total=   0.2s
[CV] C=0.2 ...........................................................
[CV] C=0.2 ...........................................................
[CV] ............................................ C=0.1, total=   0.2s
[CV] ............................................ C=0.1, total=   0.3s
[CV] C=0.2 ...........................................................
[CV] ............................................ C=0.2, total=   0.0s
[CV] C=0.2 ...........................................................
[CV] ...........

[Parallel(n_jobs=-1)]: Done  40 out of  40 | elapsed:    0.9s finished


RandomizedSearchCV(cv=4, error_score='raise',
          estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=42, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
          fit_params=None, iid=True, n_iter=10, n_jobs=-1,
          param_distributions={'C': array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ])},
          pre_dispatch='2*n_jobs', random_state=None, refit=True,
          return_train_score='warn', scoring='roc_auc', verbose=2)

In [1121]:
print(rand_search_log_reg.best_params_)

print(rand_search_log_reg.best_estimator_)

{'C': 1.0}
LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=42, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)


In [1122]:
log_reg_randomized = rand_search_log_reg.best_estimator_

y_pred_log_reg_randomized = log_reg_randomized.predict(X_train)
accuracy_score(train_labels, y_pred_log_reg_randomized)

0.8361391694725028

In [1123]:
print(confusion_matrix(train_labels, y_pred_log_reg_randomized))

print("Precision: ", precision_score(train_labels, y_pred_log_reg_randomized))

print("Recall: ", recall_score(train_labels, y_pred_log_reg_randomized))

[[487  62]
 [ 84 258]]
Precision:  0.80625
Recall:  0.7543859649122807


## KNN Classifier

In [1124]:
neigh = KNeighborsClassifier()

param_grid_neigh = {'n_neighbors' : [1, 5, 10, 20]}

grid_search_neigh = GridSearchCV(neigh, param_grid_neigh, cv = 4, scoring='roc_auc', 
                               refit = True, n_jobs = -1, verbose = 2)

grid_search_neigh.fit(X_train, train_labels)

Fitting 4 folds for each of 4 candidates, totalling 16 fits
[CV] n_neighbors=1 ...................................................
[CV] n_neighbors=1 ...................................................
[CV] n_neighbors=1 ...................................................
[CV] n_neighbors=1 ...................................................
[CV] .................................... n_neighbors=1, total=   0.1s
[CV] .................................... n_neighbors=1, total=   0.0s
[CV] n_neighbors=5 ...................................................
[CV] n_neighbors=5 ...................................................
[CV] .................................... n_neighbors=5, total=   0.0s
[CV] .................................... n_neighbors=1, total=   0.0s
[CV] n_neighbors=5 ...................................................
[CV] n_neighbors=5 ...................................................
[CV] .................................... n_neighbors=1, total=   0.1s
[CV] ............

[Parallel(n_jobs=-1)]: Done  16 out of  16 | elapsed:    0.5s finished


GridSearchCV(cv=4, error_score='raise',
       estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform'),
       fit_params=None, iid=True, n_jobs=-1,
       param_grid={'n_neighbors': [1, 5, 10, 20]}, pre_dispatch='2*n_jobs',
       refit=True, return_train_score='warn', scoring='roc_auc', verbose=2)

In [1125]:
neigh_grid = grid_search_neigh.best_estimator_

y_pred_neigh_grid = neigh_grid.predict(X_train)
accuracy_score(train_labels, y_pred_neigh_grid)

0.8305274971941639

In [1126]:
print(confusion_matrix(train_labels, y_pred_neigh_grid))

print("Precision: ", precision_score(train_labels, y_pred_neigh_grid))

print("Recall: ", recall_score(train_labels, y_pred_neigh_grid))

[[508  41]
 [110 232]]
Precision:  0.8498168498168498
Recall:  0.6783625730994152


# Model of Choice

As we can see, KNN is clearly a winner since it has the highest accuracy score. 

In [1129]:
svc_randomized = rnd_search_cv_svc.best_estimator_

y_pred_svc_rand = svc_randomized.predict(X_test)

In [1130]:
# Let's append the necessary info into the dataframe

result_test = pd.DataFrame()

In [1131]:
result_test["PassengerId"] = passenger_id_test

In [1132]:
result_test["Survived"] = y_pred_ada_rand

In [1133]:
result_test

PassengerId  Survived
0            892         0
1            893         1
2            894         0
3            895         0
4            896         1
5            897         0
6            898         1
7            899         0
8            900         1
9            901         0
10           902         0
11           903         0
12           904         1
13           905         0
14           906         1
15           907         1
16           908         0
17           909         0
18           910         1
19           911         1
20           912         0
21           913         0
22           914         1
23           915         0
24           916         1
25           917         0
26           918         1
27           919         0
28           920         0
29           921         0
..           ...       ...
388         1280         0
389         1281         0
390         1282         0
391         1283         1
392         1284         0
393         1285         0
394         1286         0
395         1287         1
396         1288         0
397         1289         1
398         1290         0
399         1291         0
400         1292         1
401         1293         0
402         1294         1
403         1295         0
404         1296         0
405         1297         0
406         1298         0
407         1299         0
408         1300         1
409         1301         1
410         1302         1
411         1303         1
412         1304         1
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         0

[418 rows x 2 columns]

In [1138]:
result_test.to_csv("result_test.csv")

# TensorFlow

# Create a Validation Set

- Take 20% of the train set as the validation set. Perfrom stratified sampling.

In [1139]:
#master_data = pd.concat([train_labels, train_data], axis = 1)

In [1140]:
#master_data

In [1141]:
#master_data.shape

In [1142]:
#from sklearn.model_selection import StratifiedShuffleSplit

In [1143]:
"""split = StratifiedShuffleSplit(n_splits = 1, test_size = 0.2, random_state = 42)
for train_index, test_index in split.split(master_data, master_data["Survived"]):
    strat_train_set = master_data.loc[train_index]
    strat_val_set = master_data.loc[test_index]"""

'split = StratifiedShuffleSplit(n_splits = 1, test_size = 0.2, random_state = 42)\nfor train_index, test_index in split.split(master_data, master_data["Survived"]):\n    strat_train_set = master_data.loc[train_index]\n    strat_val_set = master_data.loc[test_index]'

In [1144]:
rows, cols = X_train.shape

print(rows)
print(cols)


boundary = int(0.2*rows)
print(boundary)
print(rows - boundary)

891
29
178
713


In [1145]:
X_val = X_train[(891-boundary):892,] #OLD
y_val = train_labels[(891-boundary):892,] #OLD

X_train = X_train[:713,] #OLD
train_labels = train_labels[:713,] #OLD

#y_val = strat_val_set["Survived"].copy()
#X_val = strat_val_set.drop("Survived", axis = 1)

#train_labels = strat_train_set["Survived"].copy()
#X_train = strat_train_set.drop("Survived", axis = 1)

# X_train, train_labels


In [1146]:
X_train.shape

(713, 29)

In [1147]:
np.array(train_labels).shape

(713,)

In [1148]:
X_train.shape

(713, 29)

Let's do some validation set manipulation. In particular, let's reset the indexes so that they start at 0.

In [1149]:
"""#x_val_copy = X_val.reset_index()
y_val_copy = y_val.reset_index()

#x_val_copy_clean = x_val_copy.drop("index", axis = 1)
y_val_copy_clean = y_val_copy.drop("index", axis = 1)

#X_val = x_val_copy_clean
y_val = y_val_copy_clean"""

'#x_val_copy = X_val.reset_index()\ny_val_copy = y_val.reset_index()\n\n#x_val_copy_clean = x_val_copy.drop("index", axis = 1)\ny_val_copy_clean = y_val_copy.drop("index", axis = 1)\n\n#X_val = x_val_copy_clean\ny_val = y_val_copy_clean'

Apply the pipeline

In [1150]:
#X_train = preprocess_pipeline.fit_transform(X_train)

In [1151]:
#X_val = preprocess_pipeline.transform(X_val)

In [1152]:
#train_labels.value_counts()

In [1153]:
train_labels.shape

(713,)

In [1154]:
print(X_val.shape[0])

178


In [1155]:
print(X_train.shape[0])

713


In [1156]:
print(X_val.shape[0] + X_train.shape[0])

891


In [1157]:
#np.array(y_val).T[0].shape

In [1158]:
def reset_graph(seed=42):
    tf.reset_default_graph()
    tf.set_random_seed(seed)
    np.random.seed(seed)

In [1159]:
X_train.shape

(713, 29)

In [1275]:
n_inputs = 29
n_hidden1 = 110
n_hidden2 = 100
n_hidden3 = 90
n_hidden4 = 80
n_hidden5 = 80
n_hidden6 = 60
n_hidden7 = 40
n_hidden8 = 20
n_hidden9 = 20
n_hidden10 = 10
"""n_hidden11 = 100
n_hidden12 = 100
n_hidden13 = 100
n_hidden14 = 100
n_hidden15 = 100
n_hidden16 = 50
n_hidden17 = 50
n_hidden18 = 40
n_hidden19 = 20
n_hidden20 = 10"""
n_outputs = 2

In [1276]:
reset_graph()

In [1277]:
X = tf.placeholder(tf.float32, shape=(None, n_inputs), name="X")
y = tf.placeholder(tf.float32, shape=(None), name="y")

In [1278]:
# Implement dropout

training = tf.placeholder_with_default(False, shape=(), name='training') # DO NOT DELETE! ORIGINAL!

dropout_rate = 0.5  # == 1 - keep_prob # DO NOT DELETE! ORIGINAL!
X_drop = tf.layers.dropout(X, dropout_rate, training=training) # DO NOT DELETE! ORIGINAL!

with tf.name_scope("dnn"):
    hidden1 = tf.layers.dense(X_drop, n_hidden1, name="hidden1", # DO NOT DELETE! ORIGINAL!
                              activation=tf.nn.selu)             # DO NOT DELETE! ORIGINAL!
    #hidden1 = tf.layers.dense(X, n_hidden1, name="hidden1", 
    #                          activation=tf.nn.leaky_relu)
    hidden2 = tf.layers.dense(hidden1, n_hidden2, name="hidden2",
                              activation=tf.nn.selu)
    hidden3 = tf.layers.dense(hidden2, n_hidden3, name="hidden3",
                              activation=tf.nn.selu)
    hidden4 = tf.layers.dense(hidden3, n_hidden4, name="hidden4",
                              activation=tf.nn.selu)
    hidden5 = tf.layers.dense(hidden4, n_hidden5, name="hidden5",
                              activation=tf.nn.selu)
    hidden6 = tf.layers.dense(hidden5, n_hidden6, name="hidden6",
                              activation=tf.nn.selu)
    hidden7 = tf.layers.dense(hidden6, n_hidden7, name="hidden7",
                              activation=tf.nn.selu)
    hidden8 = tf.layers.dense(hidden7, n_hidden8, name="hidden8",
                              activation=tf.nn.selu)
    hidden9 = tf.layers.dense(hidden8, n_hidden9, name="hidden9",
                              activation=tf.nn.selu)
    hidden10 = tf.layers.dense(hidden9, n_hidden10, name="hidden10",
                              activation=tf.nn.selu)
    """hidden11 = tf.layers.dense(hidden10, n_hidden11, name="hidden11",
                              activation=tf.nn.selu)
    hidden12 = tf.layers.dense(hidden11, n_hidden12, name="hidden12",
                              activation=tf.nn.selu)
    hidden13 = tf.layers.dense(hidden12, n_hidden13, name="hidden13",
                              activation=tf.nn.selu)
    hidden14 = tf.layers.dense(hidden13, n_hidden14, name="hidden14",
                              activation=tf.nn.selu)
    hidden15 = tf.layers.dense(hidden14, n_hidden15, name="hidden15",
                              activation=tf.nn.selu)
    hidden16 = tf.layers.dense(hidden15, n_hidden16, name="hidden16",
                              activation=tf.nn.selu)
    hidden17 = tf.layers.dense(hidden16, n_hidden17, name="hidden17",
                              activation=tf.nn.selu)
    hidden18 = tf.layers.dense(hidden17, n_hidden18, name="hidden18",
                              activation=tf.nn.selu)
    hidden19 = tf.layers.dense(hidden18, n_hidden19, name="hidden19",
                              activation=tf.nn.selu)
    hidden20 = tf.layers.dense(hidden19, n_hidden20, name="hidden20",
                              activation=tf.nn.selu)"""
    logit = tf.layers.dense(hidden10, n_outputs, name="outputs")
    #logits = logit[:, 0] 
    logits = logit
    #logits = np.reshape(logits, (-1, 1)) # NEW!
    #logits = tf.reshape(logits, [-1, 1]) # NEW. DO NOT REMOVE!

In [1279]:
with tf.name_scope("loss"):
    #xentropy = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=y, logits=logits) # tf.transpose is new
    #xentropy = tf.nn.sigmoid_cross_entropy_with_logits(labels=y, logits=logits)
    xentropy = tf.nn.softmax_cross_entropy_with_logits(labels=tf.one_hot(tf.cast(y, tf.int32), depth = 2), logits=logits)
    loss = tf.reduce_mean(xentropy, name="loss")
    loss_summary = tf.summary.scalar('log_loss', loss)

In [1280]:
learning_rate = 0.0001

with tf.name_scope("train"):
    optimizer = tf.train.GradientDescentOptimizer(learning_rate)
    training_op = optimizer.minimize(loss)

In [1281]:
with tf.name_scope("eval"):
    correct = tf.nn.in_top_k(logits, tf.cast(y, tf.int64), 1) # tf.cast is new. 
    accuracy = tf.reduce_mean(tf.cast(correct, tf.float32))
    accuracy_summary = tf.summary.scalar('accuracy', accuracy)

In [1282]:
init = tf.global_variables_initializer()
saver = tf.train.Saver()

In [1283]:
from datetime import datetime

def log_dir(prefix=""):
    now = datetime.utcnow().strftime("%Y%m%d%H%M%S")
    root_logdir = "tf_logs"
    if prefix:
        prefix += "-"
    name = prefix + "run-" + now
    return "{}/{}/".format(root_logdir, name)

In [1284]:
logdir = log_dir("Titanic")

In [1285]:
file_writer = tf.summary.FileWriter(logdir, tf.get_default_graph())

In [1286]:
m, n = X_train.shape

In [1287]:
def random_batch(X_train, y_train, batch_size):
    rnd_indices = np.random.randint(0, len(X_train), batch_size)
#    rnd_indices = np.reshape(rnd_indices, (-1, 1))
    X_batch = X_train[rnd_indices]
    y_batch = y_train[rnd_indices] # OLD! DO NOT DELETE!
    #y_batch = np.array(y_train)
    #y_batch = y_batch[rnd_indices]
    return X_batch, y_batch

In [1288]:
n_epochs = 301
batch_size = 5
n_batches = int(np.ceil(m / batch_size))

checkpoint_path = "/tmp/2_Titanic.ckpt"
checkpoint_epoch_path = checkpoint_path + ".epoch"
final_model_path = "./2_Titanic"

best_loss = np.infty
epochs_without_progress = 0
max_epochs_without_progress = 50

with tf.Session() as sess:
    if os.path.isfile(checkpoint_epoch_path):
        # if the checkpoint file exists, restore the model and load the epoch number
        with open(checkpoint_epoch_path, "rb") as f:
            start_epoch = int(f.read())
        print("Training was interrupted. Continuing at epoch", start_epoch)
        saver.restore(sess, checkpoint_path)
    else:
        start_epoch = 0
        sess.run(init)

    for epoch in range(start_epoch, n_epochs):
        for iteration in range(n_batches):
            X_batch, y_batch = random_batch(X_train, np.array(train_labels), batch_size) # np.array is new
            #X_batch, y_batch = X_train(batch_size)
            sess.run(training_op, feed_dict={X: X_batch, y: np.array(y_batch)}) # np.array is new
        #accuracy_val, loss_val, accuracy_summary_str, loss_summary_str = sess.run([accuracy, loss, accuracy_summary, loss_summary], 
        #                                                                          feed_dict={X: X_val, y: np.array(y_val)}) # np.array is new
        # zap = np.array(fruit).flatten()
        #print(np.array(y_val).flatten()[0:10,])
        #print(np.array(y_val).T[0])
        #print(y_val)
        #tf.cast(y_val, tf.float32)
        #accuracy_val, loss_val, accuracy_summary_str, loss_summary_str = sess.run([accuracy, loss, accuracy_summary, loss_summary], feed_dict={X: X_val, y: np.array(y_val).flatten()}) # np.array is new
        #accuracy_val, loss_val, accuracy_summary_str, loss_summary_str = sess.run([accuracy, loss, accuracy_summary, loss_summary], feed_dict={X: X_val, y: np.array(y_val).T[0]}) # np.array is new. DO NOT DELETE!
        accuracy_val, loss_val, accuracy_summary_str, loss_summary_str = sess.run([accuracy, loss, accuracy_summary, loss_summary], feed_dict={X: X_val, y: np.array(y_val)}) # np.array is new

        #print(X_val.shape)
        #print(y_val.shape)
        file_writer.add_summary(accuracy_summary_str, epoch)
        file_writer.add_summary(loss_summary_str, epoch)
        if epoch % 5 == 0:
            print("Epoch:", epoch,
                  "\tValidation accuracy: {:.3f}%".format(accuracy_val * 100),
                  "\tLoss: {:.5f}".format(loss_val))
            saver.save(sess, checkpoint_path)
            with open(checkpoint_epoch_path, "wb") as f:
                f.write(b"%d" % (epoch + 1))
            if loss_val < best_loss:
                saver.save(sess, final_model_path)
                best_loss = loss_val
                os.remove(checkpoint_epoch_path)
            else:
                epochs_without_progress += 5
                if epochs_without_progress > max_epochs_without_progress:
                    print("Early stopping")
                    break


Epoch: 0 	Validation accuracy: 73.034% 	Loss: 0.54906
Epoch: 5 	Validation accuracy: 83.708% 	Loss: 0.42495
Epoch: 10 	Validation accuracy: 84.831% 	Loss: 0.39230
Epoch: 15 	Validation accuracy: 84.831% 	Loss: 0.37767
Epoch: 20 	Validation accuracy: 85.393% 	Loss: 0.36491
Epoch: 25 	Validation accuracy: 85.393% 	Loss: 0.35638
Epoch: 30 	Validation accuracy: 86.517% 	Loss: 0.35297
Epoch: 35 	Validation accuracy: 87.079% 	Loss: 0.34869
Epoch: 40 	Validation accuracy: 87.079% 	Loss: 0.34231
Epoch: 45 	Validation accuracy: 87.640% 	Loss: 0.34119
Epoch: 50 	Validation accuracy: 87.640% 	Loss: 0.33947
Epoch: 55 	Validation accuracy: 87.079% 	Loss: 0.34073
Epoch: 60 	Validation accuracy: 88.764% 	Loss: 0.33846
Epoch: 65 	Validation accuracy: 89.888% 	Loss: 0.33430
Epoch: 70 	Validation accuracy: 89.888% 	Loss: 0.33428
Epoch: 75 	Validation accuracy: 89.888% 	Loss: 0.33308
Epoch: 80 	Validation accuracy: 88.764% 	Loss: 0.33801
Epoch: 85 	Validation accuracy: 88.202% 	Loss: 0.33012
Epoch: 90 	V

Validation accuracy: Validation accuracy: 89.888% 	Loss: 0.33466

Hey, that's pretty good.

In [1289]:
os.remove(checkpoint_epoch_path)

In [1290]:
with tf.Session() as sess:
    saver.restore(sess, final_model_path)
    accuracy_value = accuracy.eval(feed_dict={X: X_train, y: train_labels})

INFO:tensorflow:Restoring parameters from ./2_Titanic


In [1291]:
accuracy_value

0.83029455

In [1269]:
with tf.Session() as sess:
    saver.restore(sess, final_model_path)
    dnn_test_output = sess.run(tf.argmax(logits,1),feed_dict={X: X_test})

INFO:tensorflow:Restoring parameters from ./2_Titanic


In [1270]:
result_test_dnn = pd.DataFrame()

In [1271]:
result_test_dnn["PassengerId"] = passenger_id_test
result_test_dnn["Survived"] = dnn_test_output

In [1272]:
result_test_dnn

PassengerId  Survived
0            892         0
1            893         0
2            894         0
3            895         0
4            896         0
5            897         0
6            898         1
7            899         0
8            900         1
9            901         0
10           902         0
11           903         0
12           904         1
13           905         0
14           906         1
15           907         1
16           908         0
17           909         0
18           910         0
19           911         1
20           912         0
21           913         1
22           914         1
23           915         0
24           916         1
25           917         0
26           918         1
27           919         0
28           920         0
29           921         0
..           ...       ...
388         1280         0
389         1281         0
390         1282         0
391         1283         1
392         1284         1
393         1285         0
394         1286         0
395         1287         1
396         1288         0
397         1289         1
398         1290         0
399         1291         0
400         1292         1
401         1293         0
402         1294         1
403         1295         0
404         1296         0
405         1297         0
406         1298         0
407         1299         0
408         1300         1
409         1301         0
410         1302         1
411         1303         1
412         1304         0
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         1

[418 rows x 2 columns]

In [1273]:
result_test_dnn.to_csv("result_test_dnn.csv")